src/machi_file_proxy.erl

```diff
@@ -729,3 +796,43 @@ maybe_regenerate_checksum(FHd, {Offset, Size, _Csum}) ->
         Error ->
             throw(Error)
     end.
+
+%% GC: make sure unwritte bytes = [{Eof, infinity}] and Eof is > max file size
+%% walk through the checksum table and make sure all chunks trimmed
+%% Then unlink the file
+-spec maybe_gc(term(), #state{}) ->
+                      {reply, term(), #state{}} | {stop, normal, term(), #state{}}.
+maybe_gc(Reply, S = #state{eof_position = Eof,
+                           max_file_size = MaxFileSize}) when Eof < MaxFileSize ->
+    lager:debug("The file is still small; not trying GC (Eof, MaxFileSize) = (~p, ~p)~n",
+                [Eof, MaxFileSize]),
+    {reply, Reply, S};
+maybe_gc(Reply, S = #state{fluname=FluName,
+                           data_filehandle = FHd,
+                           data_dir = DataDir,
+                           filename = Filename,
+                           eof_position = Eof,
+                           csum_table=CsumTable}) ->
+    case machi_csum_table:all_trimmed(CsumTable, Eof) of
+        true ->
+            lager:debug("GC? Let's do it: ~p.~n", [Filename]),
+            %% Before unlinking a file, it should inform
+            %% machi_flu_filename_mgr that this file is
+            %% deleted and mark it as "trimmed" to avoid
+            %% filename reuse and resurrection. Maybe garbage
+            %% will remain if a process crashed but it also
+            %% should be recovered at filename_mgr startup.
+
+            %% Also, this should be informed *before* file proxy
+            %% deletes files.
+            ok = machi_flu_metadata_mgr:trim_file(FluName, {file, Filename}),
+            ok = file:close(FHd),
+            {_, DPath} = machi_util:make_data_filename(DataDir, Filename),
+            ok = file:delete(DPath),
+            machi_csum_table:delete(CsumTable),
```

Comment:
I don't think the order of deletion matters, because trimmed file names are already persist (although it may need recovery..) at metadata_mgr. If any of data file or csum file remain, metadata_mgr can delete both of them according to "known_files" entries. 

Although the filename in metadata_mgr should be "trimmed_file" in current usage.